python/ccf/ledger.py

```diff
@@ -149,6 +164,200 @@ def _byte_read_safe(file, num_of_bytes):
     return ret
 
 
+class NodeStatus(IntEnum):
+    """These are the corresponding status meanings from the ccf.nodes table"""
+
+    pending = 0
+    trusted = 1
+    retired = 2
+
+
+class TxBundleInfo(NamedTuple):
+    """Bundle for transaction information required for validation """
+
+    merkle_tree: MerkleTree
+    existing_root: bytes
+    node_cert: bytes
+    signature: bytes
+    node_activity: dict
+    signing_node: int
+
+
+class LedgerValidator:
+    """
+    Ledger Validator contains the logic to verify that the ledger hasn't been tampered with.
+    It has the ability to take transactions and it maintains a MerkleTree data structure similar to CCF.
+
+    Ledger is valid and hasn't been tampered with if following conditions are met:
+        1) The merkle proof is signed by a TRUSTED node in the given network
+        2) The merkle root and signature are verified with the node cert
+        3) The merkle proof is correct for each set of transactions
+        4) The last transaction in the LedgerChunk should be that of merkle tree signing
+    """
+
+    # The node that is expected to sign the signature transaction
+    # The certificate used to sign the signature transaction
+    # https://github.com/microsoft/CCF/blob/main/src/node/nodes.h#L42
+    EXPECTED_NODE_CERT_INDEX = 1
```

Comment:
FYI, the deserialisation logic for the built-in tables will be simplified greatly once https://github.com/microsoft/CCF/issues/1993 is implemented.